design/014/ADAPTER.md

```diff
@@ -0,0 +1,127 @@
+    
+    // Cache and return - reconciler sees SecretsClient interface
+    return wrappedClient, nil
+}
+```
+
+### Server-Side Adapter
+
+The server-side adapter receives gRPC requests and translates them into `SecretsClient` interface calls. The adapter:
+
+1. Implements gRPC service interfaces (`SecretStoreProviderServer`, `GeneratorProviderServer`)
+2. Receives protobuf request messages
+3. Constructs provider instances using existing v1 provider factories
+4. Converts protobuf parameters to interface types
+5. Invokes methods on the provider's `SecretsClient` implementation
+6. Converts results to protobuf responses
+
+Provider implementations remain unchanged—they implement `ProviderInterface.NewClient()` and return `SecretsClient` instances exactly as they do for in-process use.
+
+Integration point:
+
+```go
+// providers/v2/aws/main.go (generated)
+func main() {
+    // Existing v1 provider implementation
+    v1Provider := store.NewProvider()
+    
+    // Map provider by GVK
+    providerMapping := adapterstore.ProviderMapping{
+        schema.GroupVersionKind{...}: v1Provider,
+    }
+    
+    // Adapter wraps v1 provider as gRPC server
+    adapterServer := adapter.NewServer(kubeClient, scheme, providerMapping, specMapper, generatorMapping)
+    pb.RegisterSecretStoreProviderServer(grpcServer, adapterServer)
+}
+```
+
+### Data Flow
+
+1. Controller reconciles `ExternalSecret` referencing a v2 `Provider`
+2. `clientmanager.Get()` detects v2 provider kind
+3. Manager creates client-side adapter wrapping gRPC connection
+4. Reconciler calls `client.GetSecret(ctx, ref)`
+5. Client-side adapter converts call to `pb.GetSecretRequest`
+6. gRPC request sent to remote provider server
+7. Server-side adapter receives request
+8. Server-side adapter constructs v1 provider client
+9. Server-side adapter calls `client.GetSecret(ctx, ref)` on v1 implementation
+10. Server-side adapter converts result to `pb.GetSecretResponse`
+11. Client-side adapter converts response to `[]byte`
+12. Reconciler receives secret data
+
+### Connection Management
+
+The architecture employs a global connection pool (`grpc.ConnectionPool`) to enable connection reuse across reconciliations. The `clientmanager` tracks pooled connections and releases them on `Close()`, not closing the underlying connection but returning it to the pool for subsequent use.
+
+## Consequences
+
+### Positive
+
+- **Single Codebase:** Provider implementations exist once and work both in-process and out-of-process through adapters
+- **Interface Stability:** Reconciliation logic remains unchanged; the adapter pattern is transparent
+- **Flexibility:** Providers can be deployed in-process (legacy), out-of-process (v2), or mixed
+- **Testability:** v1 provider implementations can be tested directly without gRPC infrastructure
+- **Gradual Migration:** Existing providers migrate individually without disrupting others
+
```

Comment:
```suggestion
- **New features**: This will open the door to provider implementations. Alternative deployments designs are possible, though they are not the goal of this design.
```